README.rst
Accept-Header-Match
===================

.. image:: https://travis-ci.org/vertical-knowledge/ripozo.svg?branch=master&style=flat
    :target: https://travis-ci.org/vertical-knowledge/mimetype-match
    :alt: test status

.. image:: https://readthedocs.org/projects/ripozo/badge/?version=latest
    :target: https://mimetype-match.readthedocs.org/
    :alt: Documentation Status

A tool to make parsing Accept headers and finding the appropriate mimetype to
return easy.  Simply pass your accept header and the mimetype your server can
return and it will find the best match.

There is both a programmatic interface to this package as well as a
command line interface.

Installation
------------

It can easily be installed using pip

.. code-block:: bash

    pip install mimetype-match

Or if you prefer clone the repo and install it simply run the following
from the root directory of the package.

.. code-block:: bash

    python setup.py develop

Full Documentation
------------------

`Documentation <https://mimetype-match.readthedocs.org/>`_

Using the command line interface
--------------------------------

.. code-block:: bash

    mimetype-match [accept-header] [server-mimetype1] [server-mimetype2] ...

The first argument should be the comma delimited
mimetypes as an appropriately formatted accept header.
The additional arguments are the mimetypes from the server
that you wish to match. For example you may have an Accept header
of the following:
``text/*;q=0.3, text/html;q=0.7, text/html;level=1, /html;level=2;q=0.4, */*;q=0.5``
and the server can serve the following content types:
``["text/html", "application/json", "audio/basic"]``

In this case we would run the following command:

.. code-block:: bash

    mimetype-match "text/*;q=0.3, text/html;q=0.7, text/html;level=1, /html;level=2;q=0.4, */*;q=0.5" "text/html" "application/json" "audio/basic"

This command would tell you that "text/html" is the best choice.

Using the package
-----------------

Using the package programmatically is very similar to the cli.

.. code-block:: python

    from mimetype_match import AcceptHeader

    available_types = ["text/html", "application/json", "audio/basic"]
    accept_header = AcceptHeader("text/*;q=0.3, text/html;q=0.7, text/html;level=1, /html;level=2;q=0.4, */*;q=0.5")
    best_match = accept_header.get_best_match(available_types)
    # It returns a tuple.  The first object is a MimeType object
    # which has the orinal requested type from the client
    # The second object is the available type from the
    # server that best matches the clients requested types.
    print(best_match[0].mimetype)
    print(best_match[0].weight)
    print(best_match[1])


Versioning
----------

Prior to version 1.0.0 versioning follows `sentimental
versioning <http://sentimentalversioning.org/>`_.   Releases after 1.0.0 ollow
a standard *major.minor.patch* style.

- patch: forwards and backwards compatible
- minor: backwards compatible
- major: No guarantees

Contributing
------------

Want to help out? We'd love it! Github will be the hub of development for mimetype-matching.
If you have any issues, comments, or complaints post them there.  Additionally, we
are definitely accepting pull requests (hint: we almost always love more tests and
documentation).  We do have just a few requests:

* Every method, function, and class should have a thorough docstring
* There should be at least one unit test for each function and method
* Keep your pull requests to one issue. (Preferably open an issue on github first for record keeping)


CHANGELOG.rst
CHANGELOG
`````````

1.0.5 (unreleased)
==================

- Nothing changed yet.


1.0.4 (2016-03-09)
==================

- Nothing changed yet.


1.0.3 (2016-03-09)
==================

- Nothing changed yet.


1.0.2 (2016-03-09)
==================

- Dumb package fixes


1.0.1 (2015-11-29)
==================

- Initial release


tests.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import unittest2

from mimetype_match import _is_more_specific, MimeType, AcceptHeader, InvalidMimeType,\
    get_best_match


class TestAccept(unittest2.TestCase):
    def test_is_more_specific(self):
        self.assertFalse(_is_more_specific('*/*', 'text/*'))
        self.assertFalse(_is_more_specific('*/*', '*/text'))
        self.assertFalse(_is_more_specific('text/*', 'text/something'))
        self.assertFalse(_is_more_specific('text/*', 'html/something'))

        self.assertTrue(_is_more_specific('text/*', '*/*'))
        self.assertTrue(_is_more_specific('*/text', '*/*'))
        self.assertTrue(_is_more_specific('text/something', 'text/*'))
        self.assertTrue(_is_more_specific('html/something', 'text/*'))

        self.assertIsNone(_is_more_specific('*/*', '*/*'))
        self.assertIsNone(_is_more_specific('text/*', 'html/*'))
        self.assertIsNone(_is_more_specific('*/text', '*/html'))

    def test_content_type_weight(self):
        mimes = ['text/*;q=0.3', 'text/html;level=1', 'text/html;level=2;q=0.4', '*/*;q=0.5']
        expected = [0.3, 1.0, 0.4, 0.5]
        for mime, expc in zip(mimes, expected):
            ct = MimeType(mime)
            self.assertEqual(ct.weight, expc)

    def test_content_type(self):
        mimes = ['text/*;q=0.3', 'text/html;level=1', 'text/html;level=2;q=0.4', '*/*;q=0.5']
        expected = ['text/*', 'text/html;level=1', 'text/html;level=2', '*/*']
        for mime, expc in zip(mimes, expected):
            ct = MimeType(mime)
            self.assertEqual(ct.mimetype, expc)

    def test_content_type_eq(self):
        ct = MimeType('text/*;')
        self.assertFalse(ct == None)
        self.assertFalse(ct == MimeType('text/*;q=0.5'))
        self.assertTrue(ct == MimeType('text/*;q=1'))

    def test_content_type_gt(self):
        ct = MimeType('text/*;q=0.7')
        self.assertTrue(ct > None)
        self.assertTrue(ct > MimeType('text/*;q=0.5'))
        self.assertFalse(ct > MimeType('text/*;q=1'))
        self.assertFalse(ct > MimeType('text/*;q=0.7'))

    def test_content_type_lt(self):
        ct = MimeType('text/*;q=0.7')
        self.assertFalse(ct < None)
        self.assertFalse(ct < MimeType('text/*;q=0.5'))
        self.assertTrue(ct < MimeType('text/*;q=1'))
        self.assertFalse(ct < MimeType('text/*;q=0.7'))

    def test_content_type_ge(self):
        ct = MimeType('text/*;q=0.7')
        self.assertTrue(ct >= None)
        self.assertTrue(ct >= MimeType('text/*;q=0.5'))
        self.assertFalse(ct >= MimeType('text/something;q=0.5'))
        self.assertTrue(ct >= MimeType('*/*;q=0.5'))
        self.assertFalse(ct >= MimeType('text/*;q=1'))
        self.assertTrue(ct >= MimeType('text/*;q=0.7'))

    def test_content_type_le(self):
        ct = MimeType('text/*;q=0.7')
        self.assertFalse(ct <= None)
        self.assertFalse(ct <= MimeType('text/*;q=0.5'))
        self.assertTrue(ct <= MimeType('text/something;q=0.5'))
        self.assertFalse(ct <= MimeType('*/*;q=0.5'))
        self.assertTrue(ct <= MimeType('text/*;q=1'))
        self.assertTrue(ct <= MimeType('text/*;q=0.7'))

    def test_accept_headers_order(self):
        value = ('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                 'text/html;level=2;q=0.4, */*;q=0.5')
        accept_header = AcceptHeader(value)
        content_types = [ct.mimetype for ct in accept_header]
        self.assertListEqual(['text/html;level=1', 'text/html',
                              'text/html;level=2', 'text/*', '*/*'], content_types)

    def test_accept_headers_get_match(self):
        value = ('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                 'text/html;level=2;q=0.4, */*;q=0.5')
        accept_header = AcceptHeader(value)
        best = accept_header.get_match('text/something')
        self.assertEqual(best.mimetype, 'text/*')
        best = accept_header.get_match('text/html')
        self.assertEqual(best.mimetype, 'text/html')
        best = accept_header.get_match('*/*')
        self.assertEqual(best.mimetype, '*/*')
        best = accept_header.get_match('blah/*')
        self.assertEqual(best.mimetype, '*/*')

    def test_accept_headers_get_best_match(self):
        value = ('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                 'text/html;level=2;q=0.4, */*;q=0.5')
        accept_header = AcceptHeader(value)
        best = accept_header.get_best_match(['text/json', 'blah/blah'])
        self.assertEqual(best[0].mimetype, 'text/*')
        self.assertEqual(best[1], 'text/json')

    def test_invalid_accept_header(self):
        self.assertRaises(InvalidMimeType, MimeType, '')
        self.assertRaises(InvalidMimeType, MimeType, ';  ; ;;')
        self.assertRaises(InvalidMimeType, MimeType, 'somofmdfog')
        self.assertRaises(InvalidMimeType, MimeType, 'text/html;q=blah')

    def test_accept_headers_get_match_none(self):
        value = ('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                 'text/html;level=2;q=0.4')
        accept_header = AcceptHeader(value)
        self.assertIsNone(accept_header.get_match('application/json'))

    def test_accept_headers_get_best_match_none(self):
        value = ('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                 'text/html;level=2;q=0.4')
        accept_header = AcceptHeader(value)
        self.assertIsNone(accept_header.get_best_match('application/json'))

    def test_accept_headers_no_headers(self):
        self.assertListEqual([], AcceptHeader(', ,,  , , ').mimetypes)

    def test_get_best_match(self):
        best = get_best_match(('text/*;q=0.3, text/html;q=0.7, text/html;level=1,'
                               'text/html;level=2;q=0.4, */*;q=0.5'),
                              ['text/json', 'blah/blah'])
        self.assertEqual(best[0].mimetype, 'text/*')
        self.assertEqual(best[1], 'text/json')

    def test_get_best_match_list(self):
        best = get_best_match([
            'text/*;q=0.3', 'text/html;q=0.7', 'text/html;level=1',
            'text/html;level=2;q=0.4', '*/*;q=0.5'],
            ['text/json', 'blah/blah'])
        self.assertEqual(best[0].mimetype, 'text/*')
        self.assertEqual(best[1], 'text/json')


accept_header_match.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import fnmatch
import sys

import six


class InvalidMimeType(Exception):
    """
    An exception for invalid mimetypes. Valid mimetypes
    have the format of <type>/<specific>.  They must
    can have a wildcard in either place but their must be at
    least one forward slash
    """


def _is_more_specific(first, second):
    """
    Compares how specific two mimetypes are. Returns
    True if the first is more specific, False if the second
    is more specific and None if neither is more specific.

    :param str first: The first mimetype
    :param str second: The second mimetype to compare
    :raises IndexError: Raises an index error if there
        is not at least one forward slash as required by
        mimetypes.
    """
    first = first.split('/')
    second = second.split('/')
    if first[0] == '*' and not second[0] == '*':
        return False
    if not first[0] == '*' and second[0] == '*':
        return True
    if first[1] == '*' and not second[1] == '*':
        return False
    if not first[1] == '*' and second[1] == '*':
        return True
    return None


class MimeType(object):
    """
    A Mimetype object that is comparable.  All comparison operations
    (==, >, <, etc) are available.  The comparisions are to determine
    which mimetype should be picked.  It is ordered according to
    the `Http Specification <http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html>`_.

    Additionally, it strips off the weight and makes that available
    independently and the actual mimetype as well.
    """

    def __init__(self, mimetype_string):
        """
        :param str mimetype_string: The mimetype string optionally,
            with the weight attached (e.g. "text/html;q=0.7").  Wildcard
            syntax are also valid.
        :raises InvalidMimeType: If the mimetype is not valid.
        """
        if not mimetype_string:
            raise InvalidMimeType('Invalid Mimetype')
        mimetype_string = mimetype_string.strip().strip(';')
        self._full = mimetype_string
        parts = [part.strip().strip(';') for part in mimetype_string.split(';')]
        parts = [part for part in parts if part]
        if not parts or len(parts[0].split('/')) < 2:
            raise InvalidMimeType('Invalid Mimetype')
        if len(parts) == 1 or not parts[-1].startswith('q='):
            self.mimetype = mimetype_string
            self.weight = 1.0
        else:
            try:
                self.mimetype = ';'.join(parts[:-1])
                weight = parts[-1][2:].strip()
                self.weight = float(weight)
            except ValueError:
                raise InvalidMimeType('{0} is not valid'.format(parts[-1]))

    def __eq__(self, other):
        """
        :param MimeType other: The MimeType to compare
        :return: A boolean indicating if the MimeType is equal
            equal in precidence
        :rtype: bool
        """
        if not other:
            return False
        if not self.weight == other.weight:
            return False
        specificity = _is_more_specific(self.mimetype, other.mimetype)
        return specificity is None

    def __gt__(self, other):
        """
        :param MimeType other: The MimeType to compare
        :return: A boolean indicating if the MimeType is greater than
            in precidence
        :rtype: bool
        """
        if not other:
            return True
        specificity = _is_more_specific(self.mimetype, other.mimetype)
        if specificity is not None:
            return specificity
        return self.weight > other.weight

    def __lt__(self, other):
        """
        :param MimeType other: The MimeType to compare
        :return: A boolean indicating if the MimeType is less than
            in precidence
        :rtype: bool
        """
        if not other:
            return False
        specificity = _is_more_specific(self.mimetype, other.mimetype)
        if specificity is not None:
            return not specificity
        return self.weight < other.weight

    def __ge__(self, other):
        """
        :param MimeType other: The MimeType to compare
        :return: A boolean indicating if the MimeType is greater than
            or equal in precidence
        :rtype: bool
        """
        if not other:
            return True
        specificity = _is_more_specific(self.mimetype, other.mimetype)
        if specificity is not None:
            return specificity
        return self.weight >= other.weight

    def __le__(self, other):
        """
        :param MimeType other: The MimeType to compare
        :return: A boolean indicating if the MimeType is less than
            or equal in precidence
        :rtype: bool
        """
        if not other:
            return False
        specificity = _is_more_specific(self.mimetype, other.mimetype)
        if specificity is not None:
            return not specificity
        return self.weight <= other.weight

    def is_match(self, string):
        """
        Returns true if the string matches the mimetype.
        Otherwise it returns False.

        :rtype: bool
        """
        return fnmatch.fnmatch(string, self.mimetype)


class AcceptHeader(object):
    """
    Parses an Accept header for the various mimetypes
    and sorts them in order of relevance.  It is iterable
    so it is entirely valid to simply call

    .. code-block:: python

        mimetypes = AcceptHeader(accept_header_str)
        for mimetype in mimetypes:
            # Do something

    They are ordered with the first being the preferred mimetype.
    All of the mimetypes are MimeType objects where the weight
    and actual mimetype string are available.
    """
    def __init__(self, accept_header):
        """
        :param str accept_header: The accept header with comma
            delimited mimetypes according to the
            `HTTP Specification <http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html>`_
        """
        self.mimetypes = []
        for accept in accept_header.split(','):
            accept = accept.strip()
            if not accept:
                continue
            self.mimetypes.append(MimeType(accept))
        self.mimetypes.sort(reverse=True)

    def __iter__(self):
        for mimetype in self.mimetypes:
            yield mimetype

    def get_match(self, mimetype):
        """
        Check if a mimetype
        is one of the requested mimetypes
        according to the Accept Header.

        :param str mimetype: The mimetype to compare
            to the ones in the Accept header.
        :returns: A MimeType object that includes the
            weight and mimetype string
        :rtype: MimeType
        """
        for mime in self:
            if mime.is_match(mimetype):
                return mime
        return None

    def get_best_match(self, mimetype_list):
        """
        Finds the preferred mimetype of the client
        from the mimetype list.  If none of the types
        in the mimetype list match any of the mimetypes
        provided in the Accept header, then it returns None.

        :param list[str|unicode]|tuple(unicode|str) mimetype_list: A list
            of strings to compare and find the best match
        :returns: A tuple with the first item being the matched
            mimetype from the client and the second item a string
            mimetype that the server has available.
        :rtype: tuple(MimeType, str)
        """
        content_types = [(self.get_match(mimetype), mimetype) for mimetype in mimetype_list]
        best = max(content_types)
        if not best[0]:
            return None
        return best


def get_best_match(accept_header, served_types):
    """
    A shortcut to find the best mimetype from the available mimetypes
    in the accept header.

    :param str|unicode|list[str|unicode]|tuple(str|unicode) accept_header: A comma
        delimited string that corresponds
        to the `Accept Header <http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html>`_
        or an interable of strings mimetypes.
    :param list|tuple served_types: A list of mimetypes corresponding to
    :returns: The a tuple with the first object being the matched MimeType object
        and the string mimetype that the server accepts as the second.
    :rtype: tuple(MimeType, str)|NoneType
    """
    if not isinstance(accept_header, six.string_types):
        accept_header = ','.join(accept_header)
    return AcceptHeader(accept_header).get_best_match(served_types)


def cli():
    """
    A tool to find the best matched mimetype for a particular
    Accept header. The first argument should be the comma delimited
    mimetypes as an appropriately formatted accept header.
    The additional arguments are the mimetypes from the server
    that you wish to match. For example you may have an Accept header
    of the following:
    ``text/*;q=0.3, text/html;q=0.7, text/html;level=1, /html;level=2;q=0.4, */*;q=0.5``
    and the server can serve the following content types:
    ``["text/html", "application/json", "audio/basic"]``

    In this case we would run the following command:

    .. code-block:: bash

        mimetype-match "text/*;q=0.3, text/html;q=0.7, text/html;level=1, /html;level=2;q=0.4, */*;q=0.5" "text/html" "application/json" "audio/basic"

    This command would tell you that "text/html" is the best choice.
    """
    if not sys.argv or sys.argv[0] in ['--help', '-h']:
        print(cli.__doc__)
        return
    if len(sys.argv) < 2:
        print("======================================")
        print("At least two arguments are required")
        print("======================================")
        print(cli.__doc__)
    accept = sys.argv[0]
    matches = sys.argv[1:0]
    best = get_best_match(accept, matches)
    print("The best match is to {0} with the mimetype {1}".format(best[0].content_type, best[1]))
